### ISSUE TITLE
Error `typecheck cannot infer <...>` on tilde type constraint

### ISSUE THREAD

**Issue by renom** [golangci/golangci-lint#4109]
### Welcome

- [X] Yes, I'm using a binary release within 2 latest major releases. Only such installations are supported.
- [X] Yes, I've searched similar issues on GitHub and didn't find any.
- [X] Yes, I've read the typecheck section of the FAQ (https://golangci-lint.run/usage/faq/#why-do-you-have-typecheck-errors).
- [X] Yes, I've tried with the standalone linter if available (e.g., gocritic, go vet, etc.). (https://golangci-lint.run/usage/linters/)

### Description of the problem

There's a "false alarm" error returned by golangci-lint when tilde type constraint is used in some cases (not always). I.e. the code compiles fine but typecheck returns a false-positive error.

### Version of golangci-lint

<details>

```console
golangci-lint has version 1.54.2 built with go1.21.0 from 411e0bbb on 2023-08-21T12:04:32Z
```

</details>

### Configuration

<details>

```console
go run app/main.go
```

</details>

### Go environment

<details>

```console
go version go1.21.0 linux/amd64
```

</details>

### Verbose output of running

<details>

```console
app/main.go:14:20                           typecheck  cannot infer T (app/main.go:9:18)
```

</details>

### A minimal reproducible example or link to a public repository

<details>

```go
package main

import (
	"fmt"

	"github.com/google/cel-go/common/types"
)

func IntToString[T ~int64](prefix string, value T) string {
	return prefix + fmt.Sprint(value)
}

func main() {
	var data string = IntToString("(data) ", types.Int(500))
	fmt.Println(data)
}
```

</details>

### Validation

- [X] Yes, I've included all information above (version, config, etc.).

**Comment by ldez** [comment]
Hello,

your reproducible example works as expected (no report or error):

```go
package main

import (
	"fmt"

	"github.com/google/cel-go/common/types"
)

func IntToString[T ~int64](prefix string, value T) string {
	return prefix + fmt.Sprint(value)
}

func main() {
	var data string = IntToString("(data) ", types.Int(500))
	fmt.Println(data)
}
```

```go
module github.com/golangci/sandbox

go 1.20

require github.com/google/cel-go v0.18.1

require (
	github.com/stoewer/go-strcase v1.2.0 // indirect
	google.golang.org/genproto/googleapis/api v0.0.0-20230803162519-f966b187b2e5 // indirect
	google.golang.org/genproto/googleapis/rpc v0.0.0-20230803162519-f966b187b2e5 // indirect
	google.golang.org/protobuf v1.31.0 // indirect
)
```

```console
$ golangci-lint run
$
```
